lib/3scale/api/errors.rb

```diff
@@ -0,0 +1,15 @@
+module ThreeScale
+  module API
+    class ResponseError < StandardError
+      attr_reader :response
+
+      def initialize(response)
+        @response = response
+      end
+
+      def to_s
+        response.to_s
```

Comment:
This may not be the same as https://ruby-doc.org/stdlib-2.5.3/libdoc/net/http/rdoc/Net/HTTPResponse.html#method-i-inspect. For example:

```ruby
[1] pry(main)> http = Net::HTTP.new('echo-api-invalid.3scale.net', 80)
=> #<Net::HTTP echo-api-invalid.3scale.net:80 open=false>
[2] pry(main)> response = http.get('/')
=> #<Net::HTTPMovedPermanently 301 Moved Permanently readbody=true>
[11] pry(main)> response.to_s
=> "#<Net::HTTPMovedPermanently:0x007ff1ea060688>"
[3] pry(main)> response.inspect
=> "#<Net::HTTPMovedPermanently 301 Moved Permanently readbody=true>"
```

Reply:
Right. 

The `ThreeScale::API::ResponseError` exposes the actual error to the module rescuing the error. Never pretended to wrap it or hide it. So instead of

```
def to_s
  response.inspect
end
```

maybe better to:

```
def to_s
  response.to_s
end

def inspect
  response.inspect
end
```

Just helper methods not even required to be used.

Reply:
The reason why I mentioned this is because clients of the gem may currently be rescuing from `UnexpectedResponseError`, whose `#to_s` method returned the error message that used to be passed (before this PR), i.e. it returned `response.inspect`, not `response.to_s`. Now (with this PR), no error error message is being set, which is OK because you _are_ overwriting `#to_s`; however, `#to_s` does nothing but to delegate to `@response`, thus resulting in a different message perceived by the client (in the specific case of `UnexpectedResponseError`).

I think `ForbiddenError` and `NotFoundError` behave differently. In both cases, the `response` object was being assigned as _the_ error message.

If you want to keep the exact same behaviour as before, you may need to have an optional parameter in the initializer, to override the message.

```ruby
  def initializer(response, message = response.to_s)
    super(message)
    @response = response
  end
```
And, of course, to raise that one error with `UnexpectedResponseError.new(response, response.inspect)`.

On the other hand, if you don't care about changing the error messages and really want to make all 3 types of error to behave the same, then simply pass either `response.to_s` (or `response.inspect`) to `super`.

```ruby
  def initializer(response)
    super(response.to_s) # or super(response.inspect)
    @response = response
  end
```

Either way, you don't need to override `#to_s` or `#inspect` to expose `response`, IMO.

Reply:
Let's be backward compatible :+1: 

Thanks for the tip on the error class.

Example of unexpected error with the new code I just pushed to this PR:
```ruby
[2] pry(main)> require '3scale/api'
=> true
[3] pry(main)> client = ThreeScale::API::HttpClient.new(endpoint: "http://echo-api-invalid.3scale.net", provider_key: "12345")
=> #<ThreeScale::API::HttpClient:0x0000561c7737a818
 ....
[4] pry(main)> client.get('/sdfsadf')
ThreeScale::API::HttpClient::UnexpectedResponseError: #<Net::HTTPMovedPermanently 301 Moved Permanently readbody=true>
from /home/eguzki/data/Lanak/redhat/githome/3scale-api-ruby/lib/3scale/api/http_client.rb:80:in `unexpected!'

```